From a4a801d0e96c4114267836b17b61b240de4fa202 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@freescale.com>
Date: Mon, 28 Nov 2016 13:38:34 +0800
Subject: [PATCH 1320/5242] MLK-13574-3: ARM: dts: enable audio in a seperate
 dts

commit  cd61e412026e42cbd195222d2c634b81bffcd6fa from
https://source.codeaurora.org/external/imx/linux-imx.git

Audio is in M4 domain, so we need an indepenent dts for audio.
M4 domain is controled by RTOS, this dts is for demo purpose in
linux

Signed-off-by: Shengjiu Wang <shengjiu.wang@freescale.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/boot/dts/Makefile               |    3 +-
 arch/arm/boot/dts/imx7ulp-evk-wm8960.dts |  214 ++++++++++++++++++++++++++++++
 arch/arm/boot/dts/imx7ulp-evk.dts        |   14 ++
 3 files changed, 230 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm/boot/dts/imx7ulp-evk-wm8960.dts

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index f3c070a..dfef231 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -693,7 +693,8 @@ dtb-$(CONFIG_SOC_IMX7ULP) += \
 	imx7ulp-14x14-arm2.dtb \
 	imx7ulp-evk.dtb \
 	imx7ulp-evk-emmc.dtb \
-	imx7ulp-evk-lpuart.dtb
+	imx7ulp-evk-lpuart.dtb \
+	imx7ulp-evk-wm8960.dtb
 dtb-$(CONFIG_SOC_LS1021A) += \
 	ls1021a-moxa-uc-8410a.dtb \
 	ls1021a-qds.dtb \
diff --git a/arch/arm/boot/dts/imx7ulp-evk-wm8960.dts b/arch/arm/boot/dts/imx7ulp-evk-wm8960.dts
new file mode 100644
index 0000000..9400e69
--- /dev/null
+++ b/arch/arm/boot/dts/imx7ulp-evk-wm8960.dts
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2016 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include "imx7ulp-evk.dts"
+
+/ {
+
+	aips0: aips-bus@41000000 {
+		compatible = "fsl,aips-bus", "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		reg = <0x41000000 0x80000>;
+		ranges;
+
+		pcc0: pcc0@41026000 {
+			compatible  = "fsl,imx7ulp-pcc0";
+			reg = <0x41026000 0x1000>;
+		};
+
+		clks_m4: scg0@41027000 {
+			compatible = "fsl,imx7ulp-scg0";
+			reg = <0x41027000 0x1000>;
+			clocks = <&cm4_ckil>, <&cm4_osc>, <&cm4_sirc>, <&cm4_firc>;
+			clock-names = "cm4_ckil", "cm4_osc", "cm4_sirc", "cm4_firc";
+			#clock-cells = <1>;
+		};
+
+		sai0: sai@41037000 {
+			compatible = "fsl,imx7ulp-sai";
+			reg = <0x41037000 0x1000>;
+			interrupts = <GIC_SPI 65 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&clks_m4 IMX7ULP_CM4_CLK_SAI0_IPG>,
+				 <&clks_m4 IMX7ULP_CM4_CLK_DUMMY>,
+				 <&clks_m4 IMX7ULP_CM4_CLK_SAI0_ROOT>,
+				 <&clks_m4 IMX7ULP_CM4_CLK_DUMMY>,
+				 <&clks_m4 IMX7ULP_CM4_CLK_DUMMY>;
+			clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
+			dma-names = "rx", "tx";
+			dmas = <&edma0 0 59>, <&edma0 0 60>;
+			status = "disabled";
+		};
+	};
+
+	aips1: aips-bus@41080000 {
+		compatible = "fsl,aips-bus", "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		reg = <0x41080000 0x80000>;
+		ranges;
+
+		smc0: smc0@410a4000 {
+			compatible = "fsl,imx7ulp-smc0";
+			reg = <0x410a4000 0x1000>;
+		};
+
+		sai1: sai@410AA000 {
+			compatible = "fsl,imx7ulp-sai";
+			reg = <0x410AA000 0x1000>;
+			interrupts = <GIC_SPI 66 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&clks_m4 IMX7ULP_CM4_CLK_SAI1_IPG>,
+				 <&clks_m4 IMX7ULP_CM4_CLK_DUMMY>,
+				 <&clks_m4 IMX7ULP_CM4_CLK_SAI1_ROOT>,
+				 <&clks_m4 IMX7ULP_CM4_CLK_DUMMY>,
+				 <&clks_m4 IMX7ULP_CM4_CLK_DUMMY>;
+			clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
+			dma-names = "rx", "tx";
+			dmas = <&edma0 0 61>, <&edma0 0 62>;
+			status = "disabled";
+		};
+
+		pcc1: pcc1@410b2000 {
+			compatible  = "fsl,imx7ulp-pcc1";
+			reg = <0x410b2000 0x1000>;
+		};
+	};
+
+	clocks_m4 {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cm4_ckil: clock@6 {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <32768>;
+			clock-output-names = "cm4_ckil";
+		};
+
+		cm4_osc: clock@7 {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <24000000>;
+			clock-output-names = "cm4_osc";
+		};
+
+		cm4_sirc: clock@8 {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <16000000>;
+			clock-output-names = "cm4_sirc";
+		};
+
+		cm4_firc: clock@9 {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <48000000>;
+			clock-output-names = "cm4_firc";
+		};
+	};
+
+
+	sound {
+		compatible = "fsl,imx7ulp-evk-wm8960",
+			     "fsl,imx-audio-wm8960";
+		model = "wm8960-audio";
+		cpu-dai = <&sai0>;
+		audio-codec = <&codec>;
+		codec-master;
+		/* JD3: hp detect high for headphone*/
+		hp-det = <3 0>;
+		hp-det-gpios = <&gpio3 0 0>;
+		mic-det-gpios = <&gpio3 0 0>;
+		audio-routing =
+			"Headphone Jack", "HP_L",
+			"Headphone Jack", "HP_R",
+			"Ext Spk", "SPK_LP",
+			"Ext Spk", "SPK_LN",
+			"Ext Spk", "SPK_RP",
+			"Ext Spk", "SPK_RN",
+			"LINPUT1", "Mic Jack",
+			"LINPUT3", "Mic Jack",
+			"Mic Jack", "MICB";
+
+	};
+};
+
+&iomuxc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_hog_0_1>;
+	status = "okay";
+
+	imx7ulp-evk-0 {
+		pinctrl_hog_0_1: hoggrp-0-1 {
+			fsl,pins = <
+				IMX7ULP_PAD_PTA24__PTA24		0x127
+				IMX7ULP_PAD_PTB0__CLKOUT		0x900
+			>;
+		};
+
+		pinctrl_sai0: sai0_grp {
+			fsl,pins = <
+				IMX7ULP_PAD_PTA4__I2S0_MCLK        0x700
+				IMX7ULP_PAD_PTA5__I2S0_TX_BCLK     0x0700
+				IMX7ULP_PAD_PTA2__I2S0_RXD0        0x0700
+				IMX7ULP_PAD_PTA6__I2S0_TX_FS       0x0700
+				IMX7ULP_PAD_PTA7__I2S0_TXD0        0x0700
+			>;
+		};
+	};
+};
+
+&iomuxc1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_hog_1 &pinctrl_hog_audio>;
+
+	imx7ulp-evk {
+		pinctrl_hog_audio: hoggrp-audio {
+			fsl,pins = <
+				IMX7ULP_PAD_PTF0__PTF0		0x30100
+			>;
+		};
+	};
+};
+
+&lpi2c7 {
+	status = "okay";
+
+	codec: wm8960@1a {
+		compatible = "wlf,wm8960";
+		reg = <0x1a>;
+		clocks = <&clks_m4 IMX7ULP_CLK_SCG0_CLKOUT>;
+		clock-names = "mclk";
+		wlf,shared-lrclk;
+		assigned-clocks = <&clks_m4 IMX7ULP_CM4_CLK_APLL_SEL>,
+				  <&clks_m4 IMX7ULP_CLK_SCG0_CLKOUT>;
+		assigned-clock-parents = <&clks_m4 IMX7ULP_CM4_CLK_APLL_VCO_POST_DIV2>,
+					 <&clks_m4 IMX7ULP_CM4_CLK_APLL_SEL>;
+	};
+};
+
+&clks_m4 {
+	assigned-clocks = <&clks_m4 IMX7ULP_CM4_CLK_APLL_VCO_PRE_SEL>,
+			  <&clks_m4 IMX7ULP_CM4_CLK_APLL_VCO>,
+			  <&clks_m4 IMX7ULP_CM4_CLK_APLL_VCO_POST_DIV1>,
+			  <&clks_m4 IMX7ULP_CM4_CLK_APLL_VCO_POST_DIV2>,
+			  <&clks_m4 IMX7ULP_CM4_CLK_APLL_PFD0_PRE_DIV>;
+	assigned-clock-parents = <&clks_m4 IMX7ULP_CM4_CLK_OSC>;
+	assigned-clock-rates = <0>, <540672000>, <49152000>, <12288000>, <270336000>;
+};
+
+&sai0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sai0>;
+	assigned-clocks = <&clks_m4 IMX7ULP_CM4_CLK_SAI0_SEL>,
+			  <&clks_m4 IMX7ULP_CM4_CLK_SAI0_DIV>;
+	assigned-clock-parents = <&clks_m4 IMX7ULP_CM4_CLK_APLL_PFD0_PRE_DIV>;
+	assigned-clock-rates = <0>, <12288000>;
+	fsl,dataline = <0x1 0x1>;
+	status = "okay";
+};
diff --git a/arch/arm/boot/dts/imx7ulp-evk.dts b/arch/arm/boot/dts/imx7ulp-evk.dts
index 3fb6d80..0efc09d 100644
--- a/arch/arm/boot/dts/imx7ulp-evk.dts
+++ b/arch/arm/boot/dts/imx7ulp-evk.dts
@@ -113,6 +113,13 @@
 			>;
 		};
 
+		pinctrl_lpi2c7: lpi2c7grp {
+			fsl,pins = <
+				IMX7ULP_PAD_PTF12__LPI2C7_SCL	0x527
+				IMX7ULP_PAD_PTF13__LPI2C7_SDA	0x527
+			>;
+		};
+
 		pinctrl_usdhc1: usdhc1grp {
 			fsl,pins = <
 				IMX7ULP_PAD_PTE3__SDHC1_CMD	0x843
@@ -162,6 +169,13 @@
 	};
 };
 
+&lpi2c7 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_lpi2c7>;
+};
+
 &lpuart4 { /* console */
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_lpuart4>;
-- 
1.7.9.5

